### PR TITLE
binance.createPostOnlyOrder

### PR DESCRIPTION
**Pull request by samgermain**
fixes: #15429

2022-07-04T12:52:52.936Z
Node.js: v18.4.0
CCXT v1.89.98

```
% binanceusdm createOrder ADA/USDT limit buy 23 0.464 '{"postOnly": true}'
binanceusdm.createOrder (ADA/USDT, limit, buy, 23, 0.464, [object Object])
2022-07-04T12:52:54.346Z iteration 0 passed in 224 ms

{
  info: {
    orderId: '...',
    symbol: 'ADAUSDT',
    status: 'NEW',
    clientOrderId: '...',
    price: '0.46400',
    avgPrice: '0.00000',
    origQty: '23',
    executedQty: '0',
    cumQty: '0',
    cumQuote: '0',
    timeInForce: 'GTX',
    type: 'LIMIT',
    reduceOnly: false,
    closePosition: false,
    side: 'BUY',
    positionSide: 'BOTH',
    stopPrice: '0',
    workingType: 'CONTRACT_PRICE',
    priceProtect: false,
    origType: 'LIMIT',
    updateTime: '1656939174507'
  },
  id: '...',
  clientOrderId: '...',
  timestamp: 1656939174507,
  datetime: '2022-07-04T12:52:54.507Z',
  lastTradeTimestamp: undefined,
  symbol: 'ADA/USDT',
  type: 'limit',
  timeInForce: 'PO',
  postOnly: true,
  reduceOnly: false,
  side: 'buy',
  price: 0.464,
  stopPrice: undefined,
  amount: 23,
  cost: 0,
  average: undefined,
  filled: 0,
  remaining: 23,
  status: 'open',
  fee: undefined,
  trades: [],
  fees: []
}
2022-07-04T12:52:54.346Z iteration 1 passed in 224 ms
```

```
% binance createOrder ADA/USDT limit buy 23 0.464 '{"postOnly": true}'    
binance.createOrder (ADA/USDT, limit, buy, 23, 0.464, [object Object])
2022-07-04T12:53:14.489Z iteration 0 passed in 672 ms

{
  info: {
    symbol: 'ADAUSDT',
    orderId: '...',
    orderListId: '-1',
    clientOrderId: '...',
    transactTime: '1656939194652',
    price: '0.46400000',
    origQty: '23.00000000',
    executedQty: '0.00000000',
    cummulativeQuoteQty: '0.00000000',
    status: 'NEW',
    timeInForce: 'GTC',
    type: 'LIMIT_MAKER',
    side: 'BUY'
  },
  id: '...',
  clientOrderId: '...',
  timestamp: 1656939194652,
  datetime: '2022-07-04T12:53:14.652Z',
  lastTradeTimestamp: undefined,
  symbol: 'ADA/USDT',
  type: 'limit',
  timeInForce: 'GTC',
  postOnly: true,
  reduceOnly: undefined,
  side: 'buy',
  price: 0.464,
  stopPrice: undefined,
  amount: 23,
  cost: 0,
  average: undefined,
  filled: 0,
  remaining: 23,
  status: 'open',
  fee: undefined,
  trades: [],
  fees: []
}
2022-07-04T12:53:14.489Z iteration 1 passed in 672 ms
```

### CODE REVIEW

File: js/binance.js [hunk 1]
```diff
@@ -2820,7 +2821,7 @@ module.exports = class binance extends Exchange {
         const defaultType = this.safeString2 (this.options, 'createOrder', 'defaultType', 'spot');
         const marketType = this.safeString (params, 'type', defaultType);
         const clientOrderId = this.safeString2 (params, 'newClientOrderId', 'clientOrderId');
-        const postOnly = this.safeValue (params, 'postOnly', false);
+        const postOnly = this.isPostOnly (type.toUpperCase () === 'market', false, params);
```
Comment (kroitor) [inline js/binance.js:2824]: This line looks wrong to me, an uppercase string cannot be equal to lowercase `'market'`, you probably meant this:

```suggestion
        const postOnly = this.isPostOnly (type.toUpperCase () === 'MARKET', false, params);
```

But we don't need to uppercase the type twice, a better solution would be to move line 2852 higher:

```JavaScript
        const initialUppercaseType = type.toUpperCase ();
```

**Comment by xmatthias** [comment]
In my understanding of the [timeInForce doc page](https://docs.ccxt.com/en/latest/manual.html?highlight=timeInForce#timeinforce) - we should also be able to provide `{'timeInForce': 'PO'}` as parameter to create a Post only order.

In the current CCXT version (1.92.77), that doesn't seem to work - and this PR doesn't seem to add that capability either.
instead, i get `Could not place buy order due to BadRequest. Message: binance {"code":-1115,"msg":"Invalid timeInForce."}` (with both master and this version).

I therefore don't think this PR is completely adding support for PostOnly orders.

**Comment by samgermain** [comment]
> In my understanding of the [timeInForce doc page](https://docs.ccxt.com/en/latest/manual.html?highlight=timeInForce#timeinforce) - we should also be able to provide `{'timeInForce': 'PO'}` as parameter to create a Post only order.

The implementation of `{'timeInForce': 'PO'}` as well as `{'postOnly': true}` are both handled by the method `isPostOnly` on the `exchange` class, which is called by `createOrder` on `binance`

**Comment by pietrodn** [comment]
I confirm that this fix (`timeInForce=GTX`) creates post-only orders for Binance futures (COINM and USDM).
Is there a chance to have this merged soon?

I created another PR to support `postOnly` for cross-margin: https://github.com/ccxt/ccxt/pull/14908

**Comment by Rezan-vm** [comment]
FYI, this is still an issue. Post only binanceusdm can't be made. `timeInForce=GTX` shoudl fix it.

**Pull request closed without merging**